src/silly_sprites/opengl/input.h

```diff
@@ -0,0 +1,61 @@
+#pragma once
+#include "window.hpp"
+#include "keys.hpp"
+#include "mouse.hpp"
+#include <array>
+#include <magic_enum.hpp>
+
+namespace sly::gl {
+	class Input final {
+	private:
+		static inline std::array<bool, magic_enum::enum_count<Keys>()> current_keys{  };
+		static inline std::array<bool, magic_enum::enum_count<Keys>()> last_keys{  };
+
+		static inline std::array<bool, magic_enum::enum_count<Mouse>()> current_mouse{  };
+		static inline std::array<bool, magic_enum::enum_count<Mouse>()> last_mouse{  };
+
+		static inline std::vector<std::vector<Keys>> close_input {{ 
+			{ Keys::KeyLeftAlt, Keys::KeyF4 },
+			{ Keys::KeyLeftControl, Keys::KeyLeftAlt, Keys::KeyF4 },
+			{ Keys::KeyLeftShift, Keys::KeyLeftAlt, Keys::KeyF4 },
+		}};
+
+		using mouse_pos_ty = std::pair<double, double>;
+		static inline mouse_pos_ty mouse_position{ }; 
+
+		static void update_keys(Window const& window);
+		static void update_mouse(Window const& window);
+		static void update_mouse_position(Window const& window);
+		[[nodiscard]] static bool is_quit();
+
+	public:
+		static void update(Window const& window);
+
+		[[nodiscard]] static mouse_pos_ty get_mouse_position();
+
+		// single frame
+		[[nodiscard]] static bool is_key_down(Keys key);
+		[[nodiscard]] static bool is_key_up(Keys key);
+
+		// multi-frame
+		[[nodiscard]] static bool is_key_pressed(Keys key);
+		[[nodiscard]] static bool is_key_released(Keys key);
+
+		[[nodiscard]] static bool is_key_hold(Keys key);
+		[[nodiscard]] static bool is_key_raised(Keys key);
```

Comment:
What does "single frame" and "multi-frame" mean here? I would expect that `is_key_down` returns `true` if the key is currently being pressed (either because it has been pressed this frame, or if it has been pressed before and is still held down). On the other hand, I'd expect `is_key_pressed` to only return `true` if the key has been pressed down during the current frame.
Why are there six different functions instead of only four? 🤔 `...hold` and `...raised` confuse me here.

Reply:
hey we only needed like 2h discussion to get to a version everyone could live with... 😄

**single frame:** action happens in exactly one frame so current key is either pressed or not

> down -> current
> up -> not current

**multiple frames:** action happens over two frames so 

> pressed -> not last && current
> released -> last && not current
> hold -> last && current
> raised -> not last && not current

do not give the poor purple a hard time for it, chat was on fire🔥

Reply:
He already told me that the chat was rather "demanding" today 🙈

But back to topic: What would be the use-case where you want to know if a given key is pressed **and** it was pressed during the last frame? When creating a game, there are just two cases of interest:
* was the key pressed down during this frame? (=> trigger some action)
* is the key being pressed (or held) down right now? (=> do some continuous stuff)
(same applies for released keys, of course)

I don't see a use-case where I would want to know "is the key being pressed (or held) down right now **and** was it being pressed during the last frame, too". Even if there was a use-case for that, a user could easily implement that without direct API support.

tl;dr: In my opinion, `hold` and `raised` are not needed and only lead to confusion.

Reply:
In short: that was purples stand to begin, until we confused him witch chat discussions 😄  

For what it is worth, I support you fully (given that was also my stand😈) and thankfully the ultimate power lies with you two 🙇‍♂️